### ISSUE TITLE
Problems with no-unused-vars

### ISSUE THREAD

**Issue by pigulla** [eslint/eslint#254]
Consider the following two pieces of code:
###### Program A

``` js
function f() {
    var a = [];
    return a.map();
}
```
###### Program B

``` js
function f() {
    var a = [];
    return a.map(function () {});
}
```

When linted with `no-unused-vars`, program A causes a violation ("line 2, col 8, Warning - a is defined but never used") wheres program B does not. The desired behaviour, obviously, is that neither of them should do so.

---

I've looked into this briefly and I believe the issue to be this:

When the `FunctionDeclaration` is visited, the variables in its scope are added to `variables`. However, the scope of the node is _not_ the scope created by it but the one in which itself resides.
So in the examples above, when the `FunctionDeclaration` node for `f` is visited, the scope retrieved in `populateVariables` is actually the global scope.

In the second example the same problem occurs, but this time the scope of the inner function is actually the scope of `f` so it happens to work by accident.

Also, there was an [issue](https://github.com/Constellation/escope/issues/21) with escope which didn't help :-)

---

There is also another minor issue with named function expressions:

``` js
(function g() {})();
```

This should not trigger a warning that `g` is never used.
